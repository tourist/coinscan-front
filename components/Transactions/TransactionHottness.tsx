import { useMemo } from 'react';
import { scaleLinear as d3_scaleLinear } from 'd3';
import Box from '@mui/material/Box';
import { useTheme } from '@mui/material/styles';
import { convertBigIntToNumberWithoutDecimalPlacesPrecision } from '../../utils/charts';
import settings from '../../settings.json';

const HOTNESS_MAX_OF_TOTAL_SUPPLY = 1000;

const TransactionHottness = ({ value }: { value: number }) => {
  const theme = useTheme();
  const background = theme.palette.background.default;
  const valueRounded = convertBigIntToNumberWithoutDecimalPlacesPrecision(
    BigInt(value)
  );
  const totalSupply = parseInt(settings.totalSupply);

  const color = useMemo(
    () =>
      d3_scaleLinear<string, number>()
        .domain([
          0,
          totalSupply / HOTNESS_MAX_OF_TOTAL_SUPPLY / 10,
          totalSupply / HOTNESS_MAX_OF_TOTAL_SUPPLY,
          totalSupply,
        ])
        .range([background, '#ffff00', '#ff0000', '#ff0000']),
    [background, totalSupply]
  );
  const radius = useMemo(
    () =>
      d3_scaleLinear<number, number>()
        .domain([0, totalSupply / HOTNESS_MAX_OF_TOTAL_SUPPLY, totalSupply])
        .range([10, 20, 20]),
    [totalSupply]
  );

  const radiusValue = radius(valueRounded);

  return (
    <Box
      style={{
        backgroundColor: color(valueRounded).toString(),
        width: radiusValue,
        height: radiusValue,
      }}
      sx={{
        borderRadius: '50%',
      }}
    />
  );
};

export default TransactionHottness;
