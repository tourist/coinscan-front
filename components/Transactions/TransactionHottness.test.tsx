import { render } from '@testing-library/react';
import { ThemeProvider } from '@mui/material';
import TransactionHottness from './TransactionHottness';
import { theme } from '../../utils/tests';

const TestTransactionHottnes = ({ value }: { value: number }) => (
  <ThemeProvider theme={theme}>
    <TransactionHottness value={value} />
  </ThemeProvider>
);

test('render TransactionHottness middle of scale', async () => {
  const { container } = render(
    <TestTransactionHottnes value={5000000000000} />
  );
  expect(container).toMatchSnapshot(
    'render TransactionHottness middle of scale'
  );
});

test('render TransactionHottness min of scale', async () => {
  const { container } = render(<TestTransactionHottnes value={1} />);
  expect(container).toMatchSnapshot('render TransactionHottness min of scale');
});

test('render TransactionHottnes over max of scale', async () => {
  const { container } = render(
    <TransactionHottness value={1000000000000000} />
  );
  expect(container).toMatchSnapshot(
    'render TransactionHottness over max of scale'
  );
});
