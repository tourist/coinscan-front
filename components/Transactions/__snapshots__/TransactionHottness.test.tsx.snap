// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render TransactionHottnes over max of scale: render TransactionHottness over max of scale 1`] = `
<div>
  <div
    class="MuiBox-root css-uodor8"
    style="background-color: rgb(255, 0, 0); width: 20px; height: 20px;"
  />
</div>
`;

exports[`render TransactionHottness middle of scale: render TransactionHottness middle of scale 1`] = `
<div>
  <div
    class="MuiBox-root css-uodor8"
    style="background-color: rgb(137, 137, 9); width: 10.5px; height: 10.5px;"
  />
</div>
`;

exports[`render TransactionHottness min of scale: render TransactionHottness min of scale 1`] = `
<div>
  <div
    class="MuiBox-root css-uodor8"
    style="background-color: rgb(18, 18, 18); width: 10px; height: 10px;"
  />
</div>
`;
